### ISSUE TITLE
better Map primitives for C code generation

### ISSUE THREAD

**Issue by amosr** [ambiata/icicle#324]
before we start adding other map primitives such as in #314, we might want to clean up the current map implementation (linear scan, unordered).
the thing currently worrying me about implementing better map primitives is how we'll integrate that with the melting.
at the moment we sidestep this by converting maps to arrays and loops in the flatten stage, and that works for now but it makes implementing a better map quite hard.

I'm thinking about a primitive for binary search, which takes a list of types of the input arrays..

```
PrimBinarySearch [Type]

typeOfPrim (PrimBinarySearch ts) = FunT (fmap ArrayT ts) Index
```

so for example
```
typeOfPrim (PrimBinarySearch [SumT UnitT UnitT])
= ArrayT (UnitT + UnitT) -> Index
```
will be flattened to
```
typeOfPrim (PrimBinarySearch [BoolT, UnitT, UnitT])
 = ArrayT BoolT -> ArrayT UnitT -> ArrayT UnitT -> Index
```

implementing this from the C side would probably require code generation too actually.. but we might be able to do something fancy with CPP?

I think the binary search primitive might return a pair of the index where the key should go, and a boolean saying whether it's actually there.

we would have another primitive
```
PrimArrayInsert Type
typeOfPrim (PrimArrayInsert t) = IntT -> t -> ArrayT t -> ArrayT t
```
that inserts into the array at an index, pushing everything after it back one.

**Comment by amosr** [comment]
this is no longer relevant
